lib/src/spotify_base.dart

```diff
@@ -49,46 +47,75 @@ abstract class SpotifyApiBase {
     _categories = Categories(this);
   }
 
-  Future<Null> _refreshToken() async {
-    if (_apiToken == null || _apiToken.isExpired) {
-      var headers = {'Authorization': 'Basic ${_credentials.basicAuth}'};
-      var body = {'grant_type': 'client_credentials'};
-
-      var responseJson = await _postImpl(_tokenRefreshUrl, headers, body);
-      var responseMap = json.decode(responseJson);
-
-      _apiToken = ApiToken.fromJson(responseMap);
-    }
+  SpotifyApiBase(SpotifyApiCredentials credentials,
+                 [http_all.Client httpClient]):
+        this.fromClient(clientCredentialsGrant(
+          Uri.parse(SpotifyApiBase._tokenUrl),
+          credentials.clientId, credentials.clientSecret, httpClient: httpClient));
+
+  SpotifyApiBase.fromAuthCodeGrant(AuthorizationCodeGrant grant, String responseUri) :
+      this.fromClient(grant.handleAuthorizationResponse(Uri.parse(responseUri).queryParameters));
+
+    static AuthorizationCodeGrant authorizationCodeGrant(
+      SpotifyApiCredentials credentials, http_all.BaseClient httpClient) {
+    return AuthorizationCodeGrant(
+        credentials.clientId,
+        Uri.parse(SpotifyApiBase._authorizationUrl),
+        Uri.parse(SpotifyApiBase._tokenUrl),
+        secret: credentials.clientSecret,
+        httpClient: httpClient);
   }
 
   Future<String> _get(String path) {
-    return _requestWrapper(path, (url, headers) => _getImpl(url, headers));
+    return _getImpl('${_baseUrl}/$path', const {});
```

Comment:
You're passing here `const {}` for the headers parameter - how are the headers sent to the Spotify API for authorization? like we did before with the bearer authorization?

Reply:
oauth2 Client overrides the BaseClient's send method and does that for us:
https://github.com/dart-lang/oauth2/blob/master/lib/src/client.dart#L105:
```dart
Future<http.StreamedResponse> send(http.BaseRequest request) async {
    if (credentials.isExpired) {
      if (!credentials.canRefresh) throw new ExpirationException(credentials);
      await refreshCredentials();
    }

    request.headers['authorization'] = "Bearer ${credentials.accessToken}";
...
```

Reply:
Ah slick!